packages/dev/sharedUiComponents/src/components/lines/OptionsLineComponent.tsx

```diff
@@ -0,0 +1,84 @@
+import type { ChangeEvent } from "react";
+import { useState } from "react";
+import { TextInputWithSubmit } from "../TextInputWithSubmit";
+import style from "./OptionsLineComponent.modules.scss";
+
+/**
+ * This components represents an options menu with optional
+ * customizable properties
+ */
+export interface IOption {
+    label: string;
+    value: string;
+}
+
+export interface IOptionsLineComponentProps {
+    options: IOption[];
+    addOptionPlaceholder?: string; // Placeholder text to display when adding a new option
+    onOptionAdded?: (newOption: IOption) => void; // Optional function that can be used to add a new option to the menu
+    onOptionSelected: (selectedOptionValue: string) => void;
+    selectedOptionValue: string; // The value of the currently selected option
+    validateNewOptionValue?: (newOptionValue: string) => boolean; // Optional function that can be used to validate the value of a new option
+}
+
+/**
+ * We have two possible states. The user starts in the Default option, and can choose to
+ * add a new option.
+ */
+enum OptionStates {
+    Default = 0, // Default state,
+    Adding = 1, // State when the user is adding a new option to the menu
+}
+
+const OptionAddKey = "addCustomOption";
+
+export const OptionsLineComponent = (props: IOptionsLineComponentProps) => {
+    const [optionState, setOptionState] = useState(OptionStates.Default); // State of the component
+
+    const onOptionChange = (evt: ChangeEvent<HTMLSelectElement>) => {
+        if (evt.target.value === OptionAddKey) {
+            setOptionState(OptionStates.Adding);
+        } else {
+            props.onOptionSelected(evt.target.value);
+        }
+    };
+
+    const onOptionAdd = (value: string) => {
+        const newOptionText = value;
+        const newOption = { label: newOptionText, value: newOptionText };

```

Comment:
I think there will be some problems if the user add multiple times the same text: we will have several options with the same value but they should be unique as they are used as keys.

Maybe you should generate the `value` value yourself, based on a static counter (or use a guid)?

Reply:
I didn't think of something complicated:
```typescript
        const newOption = { label: newOptionText, value: "_newoption" + (OptionsLineComponent._Counter++) };
```
and declare `_Counter` as a static:
```typescript
class OptionsLineComponent {
    ...
    private static _Counter: number = 0;
}
```

Reply:
I didn't want to change the option values, so I added a new "id" field to be the key. I also used Date.now instead of a counter to create ids since I know some of the options I use already have an id of 0, 1, etc...